### PR TITLE
KEY_PURCHASE transactions now have the lock address

### LINKED ISSUE #1834: Missing lock address on /log for some transactions
The `/log` is very useful! Unfortunately it seems to have a small bug when displaying some transactions.

As you can see some lock addresses are empty, but the link is actually there with `undefined` when pointing to etherscan: https://rinkeby.etherscan.io/address/undefined

### PR DESCRIPTION
**Pull request by cnasc**
# Description

This PR is a second, hopefully cleaner, pass at what #1707 tried to do. When we pull in transactions, any key purchase should have a field that identifies which lock it is associated with.

There is some additional work needed on the log page, however:
#1707 also tried to fix the issue of accidentally mutating the redux state. This PR does not, so that fix will have to come in a later PR.

This is now ready for review, and is much smaller and simpler than the draft PR. Still could use some advice on testing; it actually looks like the handlers aren't directly tested currently.

image

# Issues

Fixes #1834 #1704 
Refs #1707

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [x] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [x] My code follows the style guidelines of this project, including naming conventions
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: unlock-app/src/services/web3Service.js [hunk 1]
```diff
@@ -319,12 +319,18 @@ export default class Web3Service extends EventEmitter {
   parseTransactionFromInput(transactionHash, contract, input, contractAddress) {
     const transactionType = this.getTransactionType(contract, input)
 
-    this.emit('transaction.updated', transactionHash, {
+    let update = {
```
Comment (cellog) [inline unlock-app/src/services/web3Service.js]: const is fine here

File: unlock-app/src/services/web3Service.js [hunk 2]
```diff
@@ -462,14 +468,19 @@ export default class Web3Service extends EventEmitter {
       if (blockNumber - blockTransaction.blockNumber < requiredConfirmations) {
         this._watchTransaction(transactionHash)
       }
-
       // The transaction was mined, so we should have a receipt for it
-      this.emit('transaction.updated', transactionHash, {
+      let update = {
```
Comment (cellog) [inline unlock-app/src/services/web3Service.js]: and here too

File: unlock-app/src/services/web3Service.js [hunk 3]
```diff
@@ -319,12 +319,18 @@ export default class Web3Service extends EventEmitter {
...
+    let update = {
       status: 'pending',
       type: transactionType,
       confirmations: 0,
       blockNumber: Number.MAX_SAFE_INTEGER, // Asign the largest block number for sorting purposes
-    })
+    }
+    // The lock a key is associated with is the same as the contract the key
+    // purchase transaction was sent to
+    if (transactionType === TRANSACTION_TYPES.KEY_PURCHASE) {
```
Comment (julien51) [inline unlock-app/src/services/web3Service.js]: I would put that logic in the `handler` called at line 352. The code here is not specific to any type of transaction while the handler actually exists for this specific purpose ;)

File: unlock-app/src/services/web3Service.js [hunk 4]
```diff
@@ -462,14 +468,19 @@ export default class Web3Service extends EventEmitter {
...
-      this.emit('transaction.updated', transactionHash, {
+      let update = {
         status: 'mined',
         type: transactionType,
         confirmations: blockNumber - blockTransaction.blockNumber,
         blockNumber: blockTransaction.blockNumber,
-      })
+      }
+      // The lock a key is associated with is the same as the contract the key
+      // purchase transaction was sent to
```
Comment (julien51) [inline unlock-app/src/services/web3Service.js]: Same comment as above ;)

File: unlock-app/src/services/web3Service.js (hunk 3 above)
Comment (cnasc) [reply unlock-app/src/services/web3Service.js]: Sure thing! Are these the relevant methods on the contract?
- `purchaseFor`
- `purchaseForFrom`

**Pull request merged**